### PR TITLE
added atleast *args input support

### PR DESCRIPTION
**Pull request by hinriksnaer**
## Proposed changes

Following up with [atleast feature request](https://github.com/ml-explore/mlx/issues/668)

Swapped out `array` to `*args` for atleast_1d, atleast_2d, atleast_3d. 

The function now returns either an `array` or `list(array)` depending on the amount of inputs, similar to the Numpy and PyTorch implementation of the same function. The corresponding tests have also been updated.

The current implementation simply iterates over the argument list and calls the original `atleast_nd` c++ function. I struggled a bit figuring out how to get Pybind11 to output one of two different datatypes but this seems to do the trick. I also considered making a separate `std::vector<array>` output variation of the `atleast_nd` functions and have the Pybind11 simply call the correct variation depending on the input length. I wasn't able to find a previously implemented function to use as a reference so I settled on simply iterating over the arguments in the pybind11 definition using a lambda function. 

Please let me know if this breaks any conventions within the codebase or if an alternative approach is more desirable. 

## Checklist

Put an `x` in the boxes that apply.

- [x] I have read the [CONTRIBUTING](https://github.com/ml-explore/mlx/blob/main/CONTRIBUTING.md) document
- [x] I have run `pre-commit run --all-files` to format my code / installed pre-commit prior to committing changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have updated the necessary documentation (if needed)

### CODE REVIEW

**Review by awni — CHANGES_REQUESTED** [review]
Thanks for the follow up!

In the interest of keeping our C++ and Python APIs in sync, could you make this change on the C++ side and then dispatch to the appropriate function from Python?

Concretely something like:
- Have overloads in C++ for single `array` or `std::vector<array>`
- Have the second one call the first in a loop
- Dispatch in Python to the appropriate function

**Comment by hinriksnaer** [comment]
Sure thing! I will look right into that. In the interest of clarity, do you have anything in mind I could use for reference?

**Comment by awni** [comment]
I wasn't referring to anything specific above. If you run into issues implementing just post here and I'm happy to help.

File: mlx/ops.cpp [hunk 1]
```diff
@@ -3388,6 +3388,17 @@ array atleast_1d(const array& a, StreamOrDevice s /* = {} */) {
   return a;
 }
 
+std::vector<array> atleast_1d(
+    const std::vector<array>& arrays,
+    StreamOrDevice s) {
```
Comment (awni) [inline mlx/ops.cpp:3393]:
```suggestion
    StreamOrDevice s /* = {} */) {
```

File: mlx/ops.cpp [hunk 2]
```diff
@@ -3399,6 +3410,17 @@ array atleast_2d(const array& a, StreamOrDevice s /* = {} */) {
   }
 }
 
+std::vector<array> atleast_2d(
+    const std::vector<array>& arrays,
+    StreamOrDevice s) {
```
Comment (awni) [inline mlx/ops.cpp:3415]:
```suggestion
    StreamOrDevice s /* = {} */) {
```

File: mlx/ops.cpp [hunk 3]
```diff
@@ -3411,4 +3433,16 @@ array atleast_3d(const array& a, StreamOrDevice s /* = {} */) {
       return a;
   }
 }
+
+std::vector<array> atleast_3d(
+    const std::vector<array>& arrays,
+    StreamOrDevice s) {
```
Comment (awni) [inline mlx/ops.cpp:3439]:
```suggestion
    StreamOrDevice s /* = {} */) {
```

File: python/src/ops.cpp [hunk 4]
```diff
@@ -3638,62 +3638,96 @@ void init_ops(py::module_& m) {
...
-        atleast_1d(a: array, stream: Union[None, Stream, Device] = None) -> array
+        atleast_1d(*args: array, stream: Union[None, Stream, Device] = None) -> Union[array, List[array]]
 
-        Convert array to have at least one dimension.
+        Convert all arrays to have at least one dimension.
 
-        args:
-            a (array): Input array
+        Args:
+            args (array or list(array)): Input array or list of input arrays.
```
Comment (awni) [inline python/src/ops.cpp:3664]:
```suggestion
            *args: Input arrays.
```

Based on https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html

Comment (awni) [reply python/src/ops.cpp:3664]: Also would you mind changing the name from `args` to `arys` for consistency with NumPy?

File: python/src/ops.cpp [hunk 5]
```diff
@@ -3638,62 +3638,96 @@ void init_ops(py::module_& m) {
...
   m.def(
       "atleast_1d",
-      &atleast_1d,
-      "a"_a,
+      [](const py::args& args,
+         StreamOrDevice s) -> std::variant<py::object, py::list> {
+        if (args.size() == 1) {
+          py::object result;
+          result = py::cast(atleast_1d(args[0].cast<array>(), s));
+          return result;
```
Comment (awni) [inline python/src/ops.cpp:3646]:
```suggestion
          return py::cast(atleast_1d(args[0].cast<array>(), s));
```

Does that work?

File: python/src/ops.cpp [hunk 6]
```diff
@@ -3638,62 +3638,96 @@ void init_ops(py::module_& m) {
         )pbdoc");
   m.def(
       "atleast_1d",
-      &atleast_1d,
-      "a"_a,
+      [](const py::args& args,
+         StreamOrDevice s) -> std::variant<py::object, py::list> {
```
Comment (awni) [inline python/src/ops.cpp:3642]: Do you need to declare the `variant` output type here? I might just have it all be `py::object`?

File: python/src/ops.cpp [hunk 7]
```diff
@@ -3638,62 +3638,96 @@ void init_ops(py::module_& m) {
...
+          py::object result;
+          result = py::cast(atleast_1d(args[0].cast<array>(), s));
+          return result;
+        }
+
+        std::vector<array> arrays = args.cast<std::vector<array>>();
+        std::vector<array> result = atleast_1d(arrays, s);
+        py::list py_result = py::cast(result);
+
+        return py_result;
```
Comment (awni) [inline python/src/ops.cpp:3653]:
```suggestion
       return py::cast(result);
```

Does that work?

**Review by awni — CHANGES_REQUESTED** [review]
I left a few comments on `atleast_1d` which apply to the others as well. Could you check and address the ones that make sense?

File: python/src/ops.cpp (hunk 6 above)
Comment (hinriksnaer) [reply python/src/ops.cpp]: returning only py::object is sufficient, thanks for that. I thought a variant was necessary when returning a python list.

**Comment by hinriksnaer** [comment]
Here is a set of changes based on your feedback @awni. 

- [x] added /* = {} */
- [x] refactored pybind11 defenition
- [x] fixed docstring

The only thing I am uncertain about is how to rename the resulting python function to contain *arys instead of *args using pybind11.

**Comment by awni** [comment]
> The only thing I am uncertain about is how to rename the resulting python function to contain *arys instead of *args using pybind11.

I think you did that in the docstring + function signature already? That's the only change need as far as I can tell.

File: python/src/ops.cpp [hunk 8]
```diff
@@ -3638,62 +3638,72 @@ void init_ops(py::module_& m) {
...
       "atleast_3d",
-      &atleast_3d,
-      "a"_a,
+      [](const py::args& arys, StreamOrDevice s) -> py::object {
+        if (arys.size() == 1) {
+          return py::cast(atleast_3d(arys[0].cast<array>(), s));
+        }
+        return py::cast(atleast_3d(arys.cast<std::vector<array>>(), s));
+      },
       py::pos_only(),
```
Comment (awni) [inline python/src/ops.cpp:3694]: Nit: the pos_only isn't needed here.

File: python/src/ops.cpp [hunk 9]
```diff
@@ -3638,62 +3638,72 @@ void init_ops(py::module_& m) {
...
       "atleast_2d",
-      &atleast_2d,
-      "a"_a,
+      [](const py::args& arys, StreamOrDevice s) -> py::object {
+        if (arys.size() == 1) {
+          return py::cast(atleast_2d(arys[0].cast<array>(), s));
+        }
+        return py::cast(atleast_2d(arys.cast<std::vector<array>>(), s));
+      },
       py::pos_only(),
```
Comment (awni) [inline python/src/ops.cpp:3670]: Nit: the pos_only isn't needed here.

File: python/src/ops.cpp [hunk 10]
```diff
@@ -3638,62 +3638,72 @@ void init_ops(py::module_& m) {
...
       "atleast_1d",
-      &atleast_1d,
-      "a"_a,
+      [](const py::args& arys, StreamOrDevice s) -> py::object {
+        if (arys.size() == 1) {
+          return py::cast(atleast_1d(arys[0].cast<array>(), s));
+        }
+        return py::cast(atleast_1d(arys.cast<std::vector<array>>(), s));
+      },
       py::pos_only(),
```
Comment (awni) [inline python/src/ops.cpp:3647]: Nit: the pos_only isn't needed here.

**Comment by hinriksnaer** [comment]
Pushed the changes [here](https://github.com/hinriksnaer/mlx/tree/array_atleast).

Hopefully they made it to the push request.

**Review by awni — APPROVED** [review]
LGTM, thanks!!

**Pull request merged**